src/test/java/com/puppycrawl/tools/checkstyle/JavaAstVisitorTest.java

```diff
@@ -190,4 +204,34 @@ public void testNullSelfInAddLastSibling() throws Exception {
                 .that(addLastSibling.invoke(JavaAstVisitor.class, null, null))
                 .isNull();
     }
+
+    @Test
+    public void testNoStackOverflowOnDeepStringConcat() throws Exception {
+        final File file =
+                new File(getPath("InputJavaAstVisitorNoStackOverflowOnDeepStringConcat.java"));
+        final FileText fileText = new FileText(file, StandardCharsets.UTF_8.name());
+        final FileContents contents = new FileContents(fileText);
+
+        final String fullText = contents.getText().getFullText().toString();
+        final CharStream codePointCharStream = CharStreams.fromString(fullText);
+        final JavaLanguageLexer lexer = new JavaLanguageLexer(codePointCharStream, true);
+        lexer.setCommentListener(contents);
+
+        final CommonTokenStream tokenStream = new CommonTokenStream(lexer);
+        final JavaLanguageParser parser = new JavaLanguageParser(tokenStream);
+
+        final JavaLanguageParser.CompilationUnitContext compilationUnit = parser.compilationUnit();
+
+        // We restrict execution to use limited resources here, so that we can
+        // kill surviving pitest mutation from removal of nested binary operation
+        // optimization in JavaAstVisitor#visitBinOp. Limited resources (small stack size)
+        // ensure that we throw a StackOverflowError if optimization is removed.
```

Comment:
Left a comment here to explain why we have this test; I should have done this in the original test, but did not.